### PR TITLE
Using a DbSession and an EngineSingleton to inject the engine

### PR DESCRIPTION
**Pull request by josvandervelde**
Small refactor to make the code a bit cleaner.

Using a DbSession and an EngineSingleton to inject the engine, instead of giving it to every router. This is cleaner (I think) and makes it easy to patch the engine for unittests. Note that FastAPI Depends would work as well for the routers, but not for other parts of the code.

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -100,11 +100,9 @@ def create_app() -> FastAPI:
             "scopes": KEYCLOAK_CONFIG.get("scopes"),
         },
     )
-    engine = sqlmodel_engine(args.rebuild_db)
-    with engine.connect() as connection:
-        AIoDConcept.metadata.create_all(connection, checkfirst=True)
-        connection.commit()
-    with Session(engine) as session:
+    drop_or_create_database(delete_first=args.rebuild_db == "allways")
```
Comment (Taniya-Das) [inline src/main.py:103]: typo "always"

File: src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py [hunk 2]
```diff
@@ -66,7 +64,7 @@ def set_up(
     specified endpoint for the given resource.
```
Comment (Taniya-Das) [inline src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py:64]: remove 'engine: Engine' (line 54)

File: src/tests/routers/generic/test_router_deprecation.py [hunk 3]
```diff
@@ -39,7 +39,7 @@ def test_deprecated_router(
 ):
```
Comment (Taniya-Das) [inline src/tests/routers/generic/test_router_deprecation.py:39]: remove 'engine_test_resource_filled: Engine' from function parameter list.

File: src/tests/uploader/huggingface/test_dataset_uploader.py [hunk 4]
```diff
@@ -53,11 +48,10 @@ def test_happy_path_new_repository(
     assert id_response == 1
 
 
-@pytest.mark.skip(reason="We'll fix this in a separate PR")
 def test_repo_already_exists(client: TestClient, engine: Engine, mocked_privileged_token: Mock):
```
Comment (Taniya-Das) [inline src/tests/uploader/huggingface/test_dataset_uploader.py:51]: remove 'engine: Engine'

**Review by Taniya-Das — CHANGES_REQUESTED** [review]
small changes

File: src/main.py (hunk 1 above)
Comment (josvandervelde) [reply src/main.py]: Fixed: https://github.com/aiondemand/AIOD-rest-api/pull/199/commits/6b3e06761a6584b4f562ef7fc4a8668e54e88008

File: src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py (hunk 2 above)
Comment (josvandervelde) [reply src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py:62]: Fixed: https://github.com/aiondemand/AIOD-rest-api/pull/199/commits/47b254c7e5d729f8767fbb1f45020f69660ae5e7

File: src/tests/uploader/huggingface/test_dataset_uploader.py [hunk 5]
```diff
@@ -1,27 +1,22 @@
...
+from database.session import DbSession
 from tests.testutils.paths import path_test_resources
 
 
-@pytest.mark.skip(reason="We'll fix this in a separate PR")
-# TODO: there are errors when running these tests: "... is not bound to a Session; lazy load
-#  operation of attribute 'license' cannot proceed".
-#  See TODOs at hugging_face_uploader.py.
 def test_happy_path_new_repository(
     client: TestClient, engine: Engine, mocked_privileged_token: Mock, dataset: Dataset
```
[the remark this answers is not in the document]
Comment (josvandervelde) [reply src/tests/uploader/huggingface/test_dataset_uploader.py]: Fixed: https://github.com/aiondemand/AIOD-rest-api/pull/199/commits/47b254c7e5d729f8767fbb1f45020f69660ae5e7

File: src/tests/uploader/huggingface/test_dataset_uploader.py (hunk 4 above)
Comment (josvandervelde) [reply src/tests/uploader/huggingface/test_dataset_uploader.py]: Fixed: https://github.com/aiondemand/AIOD-rest-api/pull/199/commits/47b254c7e5d729f8767fbb1f45020f69660ae5e7

**Comment by josvandervelde** [comment]
> I am not sure why in some tests sqlalchemy.engine.Engine is still used.

Thanks for the review! I indeed forgot to remove many usages of the `Engine` in the test. Fixed it in https://github.com/aiondemand/AIOD-rest-api/pull/199/commits/47b254c7e5d729f8767fbb1f45020f69660ae5e7.

**Review by Taniya-Das — APPROVED** [review]
Looks good to me.

**Pull request merged**